rskj-core/src/main/java/co/rsk/validators/ForkDetectionDataRule.java

```diff
@@ -0,0 +1,133 @@
+        this.mainchainView = mainchainView;
+        this.forkDetectionDataCalculator = forkDetectionDataCalculator;
+        this.requiredBlocksForForkDetectionDataCalculation = requiredBlocksForForkDetectionDataCalculation;
+    }
+
+    @Override
+    public boolean isValid(Block block) {
+        return isValid(block.getHeader());
+    }
+
+    @Override
+    public boolean isValid(BlockHeader header) {
+        if(activationConfig.isActive(ConsensusRule.RSKIP110, header.getNumber()) )
+        {
+            if (hasForkDetectionDataWhereItShouldNotHave(header)){
+                return false;
+            }
+
+            if(requiredBlocksForForkDetectionDataCalculation < header.getNumber()) {
+                List<BlockHeader> headersView = mainchainView.get(
+                        header.getParentHash(),
+                        requiredBlocksForForkDetectionDataCalculation
+                );
+                if (!hasEnoughBlocksToCalculateForkDetectionData(header, headersView)) {
+                    return false;
+                }
+
+                return isForkDetectionDataEqual(header, headersView);
+            }
+        }
+
+        return true;
+    }
+
+    private boolean hasForkDetectionDataWhereItShouldNotHave(BlockHeader header) {
+        if(header.getNumber() < requiredBlocksForForkDetectionDataCalculation
+                && header.getMiningForkDetectionData().length > 0) {
+            logger.warn("Header for block number {} includes fork detection data but it should not.",
+                    header.getNumber());
+            panicProcessor.panic(
+                    PANIC_LOGGING_TOPIC,
+                    "Header includes fork detection data but it should not."
+            );
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isForkDetectionDataEqual(BlockHeader header, List<BlockHeader> headersView) {
+        byte[] expectedForkDetectionData = forkDetectionDataCalculator.calculateWithBlockHeaders(headersView);
+        if (!Arrays.equals(expectedForkDetectionData, header.getMiningForkDetectionData())) {
+            logger.warn("Fork detection data does not match. Expected {} Actual: {}",
+                    ByteUtil.toHexString(expectedForkDetectionData),
+                    ByteUtil.toHexString(header.getMiningForkDetectionData())
+            );
+            panicProcessor.panic(
+                    PANIC_LOGGING_TOPIC,
+                    "Block hash for merged mining does not match."
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean hasEnoughBlocksToCalculateForkDetectionData(BlockHeader header, List<BlockHeader> headersView) {
+        if (headersView.size() != requiredBlocksForForkDetectionDataCalculation) {
+            logger.error("Missing blocks to calculate fork detection data. Block hash {} ", header.getShortHash());
+            panicProcessor.panic(
```

Comment:
This would be a completely abnormal situation. Shouldn't it be better to just shutdown the node?

Reply:
I agree it's an abnormal situation but I'm not sure if we are handling this kind of cases with a shutdown.
Let me check that and I'll make the change if necessary.